llvm/lib/Target/AIE/AIEEliminateDuplicatePHI.cpp

```diff
@@ -0,0 +1,161 @@
+    TargetPassConfig &TPC = getAnalysis<TargetPassConfig>();
+    // Enable CSE to set up GISelObserverWrapper object
+    GISelCSEAnalysisWrapper &Wrapper =
+        getAnalysis<GISelCSEAnalysisWrapperPass>().getCSEWrapper();
+    auto *CSEInfo = &Wrapper.get(TPC.getCSEConfig());
+    std::unique_ptr<MachineIRBuilder> Builder =
+        CSEInfo ? std::make_unique<CSEMIRBuilder>()
+                : std::make_unique<MachineIRBuilder>();
+    Builder->setMF(MF);
+    MachineIRBuilder &MIB = *Builder;
+    // Set Observer to inform change of register type or change in op-code
+    GISelObserverWrapper Observer;
+    if (CSEInfo) {
+      Observer.addObserver(CSEInfo);
+      MIB.setChangeObserver(Observer);
+    }
+
+    bool Changed = false;
+    for (MachineBasicBlock &MBB : MF) {
+      // If a BB has PHI nodes, the very first instruction will be PHI
+      if (MBB.empty() ||
+          MBB.begin()->getOpcode() != llvm::TargetOpcode::G_PHI) {
+        continue;
+      }
+      Changed = processPHI(MBB, MIB, Observer);
+    }
+    return Changed;
+  }
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.addRequired<MachineModuleInfoWrapperPass>();
+    AU.addRequired<GISelCSEAnalysisWrapperPass>();
+    AU.addRequired<TargetPassConfig>();
+    AU.setPreservesAll();
+  }
+
+  StringRef getPassName() const override { return AIE_ELIMINATE_DUPLICATE_PHI; }
+
+private:
+  bool eliminateDuplicatePHI(
+      std::set<std::pair<MachineInstr *, MachineInstr *>> &IdenticalPHI,
+      MachineIRBuilder &MIB, GISelObserverWrapper &Observer) {
+    bool Changed = false;
+    if (!IdenticalPHI.empty()) {
+      LLVM_DEBUG(dbgs() << "Identical PHI nodes found\n");
+      for (auto &[BasePHI, DuplicatePHI] : IdenticalPHI) {
+        LLVM_DEBUG(dbgs() << "Base PHI: " << *BasePHI << "\n");
+        LLVM_DEBUG(dbgs() << "Duplicate PHI: " << *DuplicatePHI << "\n");
+        MIB.setInstr(*DuplicatePHI);
+        MIB.buildCopy(DuplicatePHI->getOperand(0).getReg(),
+                      BasePHI->getOperand(0).getReg());
+        Observer.changingInstr(*DuplicatePHI);
+        Observer.changedInstr(*DuplicatePHI);
+        DuplicatePHI->eraseFromParent();
+        Changed = true;
+      }
+    }
+    return Changed;
+  }
+
+  bool processPHI(MachineBasicBlock &MBB, MachineIRBuilder &MIB,
+                  GISelObserverWrapper &Observer) {
+
+    // BB and Virtual Register pair
+    using BBVRegPair = std::pair<unsigned, unsigned>;
+    struct BBVRegPairSetHash {
+      size_t operator()(const std::set<BBVRegPair> &s) const {
+        size_t hash = 0;
+        for (const auto &[BBNumber, VRegNumber] : s) {
+          hash ^= std::hash<unsigned>()(BBNumber) ^
+                  std::hash<unsigned>()(VRegNumber);
+        }
+        return hash;
+      }
+    };
+    std::unordered_map<std::set<BBVRegPair>, MachineInstr *, BBVRegPairSetHash>
+        PHIList;
+    std::set<std::pair<MachineInstr *, MachineInstr *>> IdenticalPHI;
```

Comment:
That seems quite complicated. I'd rather use `MachineInstr::isIdenticalTo()` to be really safe, and maybe have an easier algorithm which is easy to verify.

E.g. For each `G_PHI` in a BB, verify if it `isIdenticalTo` a `G_PHI` that precedes it. That's quadratic, but we can skip phis that's aren't of `p0` type, so honestly I don't think it matters much.

Reply:
We can then immediately eliminate redundant phis, no need to keep state and do the elimination in a second step.

Reply:
The approach has been inspired from a PHI elimination pass which works at IR level and not on MIR

1. Use of `MachineInstr::isIdenticalTo()` is not safe since it does not check for commutative property.  Kindly refer to `llvm/test/CodeGen/AIE/GlobalISel/eliminate-duplicate-phi.mir`
2. Modifying/Adding instruction while iterating leads to segfaults.

Based on the above review have update the pass to consider G_PHI with pointers having default address space.

Reply:
1. Do we need to know about commutative operands? I would expect all of the PHI operands to be in the same order, that is probably derived from the order of the BB predecessors.
2. I think we just need `make_early_inc_range`, this is how most passes deal with deleting instructions safely.

Reply:
The need to handle commutative property for PHI nodes come an issue that was fixed by https://github.com/llvm/llvm-project/pull/75765

The idea of using hash map come from https://github.com/llvm/llvm-project/commit/aadf55d1cea24a4e5384ab8546c3d794cb1ec724 and how it affects the compile time https://llvm-compile-time-tracker.com/compare.php?from=5f4e9bf6416e45eba483a4e5e263749989fdb3b3&to=4739e6e4eb54d3736e6457249c0919b30f6c855a&stat=instructions  given we are striving for minimal compile, I prefer using the hash map.

Reply:
Creating the hash maps can also be expensive, I'm not sure we'll see any gain for our use case. Looking at the commit you mention, they only use the set approach when there are more than 32 PHI nodes. That won't be the case for us if we only look at `p0` PHI nodes. Again, I'd rather have simple and concise code that's easy to maintain unless we can really motivate the compile time gains.

Reply:
have updated the code to remove duplicate PHI on use, and not using Hash maps any more.